In [133]:
import numpy as np
import pandas as pd

In [134]:
df = pd.read_excel("Bank_Personal_Loan_Modelling.xlsx", sheet_name="Bank_Personal_Loan_Modelling");
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
ID                       5000 non-null int64
Age (in years)           5000 non-null int64
Experience (in years)    5000 non-null int64
Income (in K/month)      5000 non-null int64
ZIP Code                 5000 non-null int64
Family members           4982 non-null float64
CCAvg                    5000 non-null float64
Education                5000 non-null int64
Mortgage                 5000 non-null int64
Personal Loan            5000 non-null int64
Securities Account       5000 non-null int64
CD Account               5000 non-null int64
Online                   5000 non-null int64
CreditCard               5000 non-null int64
dtypes: float64(2), int64(12)
memory usage: 547.0 KB


In [135]:
df.head()

ID  Age (in years)  Experience (in years)  Income (in K/month)  ZIP Code  \
0   1              25                      1                   49     91107   
1   2              45                     19                   34     90089   
2   3              39                     15                   11     94720   
3   4              35                      9                  100     94112   
4   5              35                      8                   45     91330   

   Family members  CCAvg  Education  Mortgage  Personal Loan  \
0             4.0    1.6          1         0              0   
1             3.0    1.5          1         0              0   
2             1.0    1.0          1         0              0   
3             1.0    2.7          2         0              0   
4             4.0    1.0          2         0              0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [136]:
df[df.isnull().any(axis=1)]

ID  Age (in years)  Experience (in years)  Income (in K/month)  \
20      21              56                     31                   25   
58      59              28                      2                   93   
98      99              49                     23                   94   
161    162              61                     35                   80   
235    236              38                      8                   71   
289    290              42                     15                   24   
487    488              39                     13                   88   
721    722              49                     24                   39   
1460  1461              40                     16                   85   
1461  1462              54                     28                   48   
2399  2400              62                     36                   41   
2832  2833              45                     21                  133   
3701  3702              58                     33                   95   
4135  4136              48                     23                  168   
4138  4139              47                     22                  114   
4402  4403              55                     25                   52   
4403  4404              50                     24                  112   
4763  4764              51                     25                  173   

      ZIP Code  Family members  CCAvg  Education  Mortgage  Personal Loan  \
20       94015             NaN    0.9          2       111              0   
58       94065             NaN    0.2          1         0              0   
98       92374             NaN    0.3          1         0              0   
161      95053             NaN    2.8          1         0              0   
235      94720             NaN    1.8          3         0              0   
289      92121             NaN    1.0          2         0              0   
487      94117             NaN    1.4          2         0              0   
721      92717             NaN    1.4          3         0              0   
1460     92677             NaN    0.2          3         0              0   
1461     93022             NaN    0.2          1         0              0   
2399     90245             NaN    1.0          3       154              0   
2832     92056             NaN    5.7          3         0              1   
3701     90503             NaN    2.6          1         0              0   
4135     95929             NaN    2.8          1       308              0   
4138     95819             NaN    0.6          1         0              0   
4402     90095             NaN    1.4          3       207              0   
4403     92064             NaN    0.0          1         0              0   
4763     95051             NaN    0.5          2         0              1   

      Securities Account  CD Account  Online  CreditCard  
20                     0           0       1           0  
58                     0           0       0           0  
98                     0           0       1           0  
161                    0           0       1           0  
235                    0           0       1           0  
289                    0           0       1           1  
487                    0           0       1           1  
721                    0           0       1           0  
1460                   0           0       1           1  
1461                   0           0       1           0  
2399                   1           0       1           0  
2832                   0           1       1           1  
3701                   0           0       1           0  
4135                   0           0       1           0  
4138                   1           1       1           1  
4402                   1           0       0           0  
4403                   0           0       0           0  
4763                   0           0       1           0

In [137]:
df[df.isna().any(axis=1)]

ID  Age (in years)  Experience (in years)  Income (in K/month)  \
20      21              56                     31                   25   
58      59              28                      2                   93   
98      99              49                     23                   94   
161    162              61                     35                   80   
235    236              38                      8                   71   
289    290              42                     15                   24   
487    488              39                     13                   88   
721    722              49                     24                   39   
1460  1461              40                     16                   85   
1461  1462              54                     28                   48   
2399  2400              62                     36                   41   
2832  2833              45                     21                  133   
3701  3702              58                     33                   95   
4135  4136              48                     23                  168   
4138  4139              47                     22                  114   
4402  4403              55                     25                   52   
4403  4404              50                     24                  112   
4763  4764              51                     25                  173   

      ZIP Code  Family members  CCAvg  Education  Mortgage  Personal Loan  \
20       94015             NaN    0.9          2       111              0   
58       94065             NaN    0.2          1         0              0   
98       92374             NaN    0.3          1         0              0   
161      95053             NaN    2.8          1         0              0   
235      94720             NaN    1.8          3         0              0   
289      92121             NaN    1.0          2         0              0   
487      94117             NaN    1.4          2         0              0   
721      92717             NaN    1.4          3         0              0   
1460     92677             NaN    0.2          3         0              0   
1461     93022             NaN    0.2          1         0              0   
2399     90245             NaN    1.0          3       154              0   
2832     92056             NaN    5.7          3         0              1   
3701     90503             NaN    2.6          1         0              0   
4135     95929             NaN    2.8          1       308              0   
4138     95819             NaN    0.6          1         0              0   
4402     90095             NaN    1.4          3       207              0   
4403     92064             NaN    0.0          1         0              0   
4763     95051             NaN    0.5          2         0              1   

      Securities Account  CD Account  Online  CreditCard  
20                     0           0       1           0  
58                     0           0       0           0  
98                     0           0       1           0  
161                    0           0       1           0  
235                    0           0       1           0  
289                    0           0       1           1  
487                    0           0       1           1  
721                    0           0       1           0  
1460                   0           0       1           1  
1461                   0           0       1           0  
2399                   1           0       1           0  
2832                   0           1       1           1  
3701                   0           0       1           0  
4135                   0           0       1           0  
4138                   1           1       1           1  
4402                   1           0       0           0  
4403                   0           0       0           0  
4763                   0           0       1           0

In [138]:
#Replacing NaN values with 0
df.fillna(value = 0, axis = 1, inplace = True)

In [139]:
#Capping Outliers in the given dataset. We shall only consider features which shall qualify for outlier analysis i.e.
#features with continuous values and not features that are categorical, order or count variables as they are ineligible for outlier
#analysis. Features which qualify for outlier analysis are Age, Experience, Income, Family Members, CCAvg and Mortgage
#in 'toCapDf' below.
capDf = pd.DataFrame()
toCapDf = df.iloc[:, [1, 2, 3, 5, 6, 8]]
for col in df.columns:
    if(df[col].dtype == object):
        capDf[col] = df[col]
        continue
    if(col in toCapDf.columns):
        percentiles = toCapDf[col].quantile([0.25,0.75]).values
        Q1 = percentiles[0]
        Q3 = percentiles[1]
        IQR = Q3 - Q1
        minCap = Q1 - (IQR) * 1.5
        maxCap = Q3 + (IQR) * 1.5
        capDf[col] = toCapDf[col][(toCapDf[col] >= minCap) & (toCapDf[col] <= maxCap)]
    else:
        capDf[col] = df[col]
    
capDf.dropna(inplace=True);

In [140]:
#Printing the outliers which were removed (In all, 602 eligible outlier rows were removed)
outlierDf = pd.concat([df, capDf])
outlierDf = outlierDf.reset_index(drop=True)
df_gpby = outlierDf.groupby(list(outlierDf.columns))
idx = [x[0] for x in df_gpby.groups.values() if len(x) == 1]
outlierDf.reindex(idx)

ID  Age (in years)  Experience (in years)  Income (in K/month)  \
9       10              34                      9                180.0   
18      19              46                     21                193.0   
22      23              29                      5                 62.0   
39      40              38                     13                 80.0   
42      43              32                      7                132.0   
44      45              46                     20                104.0   
47      48              37                     12                194.0   
53      54              50                     26                190.0   
55      56              41                     17                139.0   
59      60              31                      5                188.0   
61      62              47                     21                125.0   
66      67              62                     36                105.0   
90      91              55                     30                118.0   
112    113              40                     15                 82.0   
119    120              32                      7                112.0   
130    131              28                      4                 81.0   
131    132              58                     34                149.0   
145    146              59                     35                124.0   
150    151              46                     22                118.0   
151    152              26                      0                132.0   
160    161              29                      0                134.0   
172    173              38                     13                171.0   
174    175              42                     17                168.0   
187    188              46                     21                159.0   
194    195              53                     29                144.0   
209    210              64                     39                172.0   
213    214              57                     33                155.0   
227    228              47                     23                148.0   
242    243              41                     16                 75.0   
243    244              65                     39                170.0   
...    ...             ...                    ...                  ...   
4672  4673              52                     26                180.0   
4674  4675              40                     14                 93.0   
4678  4679              33                      7                115.0   
4679  4680              26                      0                161.0   
4698  4699              48                     22                162.0   
4700  4701              31                      7                170.0   
4714  4715              27                      3                 81.0   
4719  4720              32                      8                140.0   
4735  4736              34                      9                 84.0   
4755  4756              59                     35                151.0   
4765  4766              58                     34                 82.0   
4782  4783              26                      0                150.0   
4812  4813              29                      4                184.0   
4842  4843              49                     23                174.0   
4844  4845              31                      6                 81.0   
4847  4848              37                     11                 65.0   
4859  4860              34                      8                165.0   
4865  4866              50                     24                133.0   
4873  4874              59                     35                165.0   
4888  4889              25                      1                121.0   
4895  4896              45                     20                201.0   
4899  4900              54                     29                 85.0   
4908  4909  

In [141]:
capDf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4398 entries, 0 to 4999
Data columns (total 14 columns):
ID                       4398 non-null int64
Age (in years)           4398 non-null int64
Experience (in years)    4398 non-null int64
Income (in K/month)      4398 non-null float64
ZIP Code                 4398 non-null int64
Family members           4398 non-null float64
CCAvg                    4398 non-null float64
Education                4398 non-null int64
Mortgage                 4398 non-null float64
Personal Loan            4398 non-null int64
Securities Account       4398 non-null int64
CD Account               4398 non-null int64
Online                   4398 non-null int64
CreditCard               4398 non-null int64
dtypes: float64(4), int64(10)
memory usage: 515.4 KB


In [142]:
capDf.corr()

ID  Age (in years)  Experience (in years)  \
ID                     1.000000       -0.004234              -0.003173   
Age (in years)        -0.004234        1.000000               0.994115   
Experience (in years) -0.003173        0.994115               1.000000   
Income (in K/month)   -0.009554       -0.035804              -0.028003   
ZIP Code               0.008102       -0.030073              -0.028941   
Family members        -0.016114       -0.059064              -0.063533   
CCAvg                 -0.003565       -0.023779              -0.026405   
Education              0.023899        0.034672               0.006740   
Mortgage              -0.017260       -0.007577              -0.006866   
Personal Loan         -0.021774       -0.015132              -0.011917   
Securities Account    -0.019406        0.005677               0.004286   
CD Account            -0.008217        0.005973               0.008487   
Online                -0.008090        0.018553               0.018213   
CreditCard             0.016780        0.011078               0.012537   

                       Income (in K/month)  ZIP Code  Family members  \
ID                               -0.009554  0.008102       -0.016114   
Age (in years)                   -0.035804 -0.030073       -0.059064   
Experience (in years)            -0.028003 -0.028941       -0.063533   
Income (in K/month)               1.000000 -0.014084       -0.138600   
ZIP Code                         -0.014084  1.000000        0.007098   
Family members                   -0.138600  0.007098        1.000000   
CCAvg                             0.518550 -0.008175       -0.054160   
Education                        -0.149095 -0.024126        0.022929   
Mortgage                         -0.056477  0.010120        0.027610   
Personal Loan                     0.466305  0.002841        0.013074   
Securities Account                0.001725  0.000313        0.022259   
CD Account                        0.114803  0.016307       -0.000306   
Online                            0.004190  0.023776        0.011020   
CreditCard                        0.004247  0.002531        0.000842   

                          CCAvg  Education  Mortgage  Personal Loan  \
ID                    -0.003565   0.023899 -0.017260      -0.021774   
Age (in years)        -0.023779   0.034672 -0.007577      -0.015132   
Experience (in years) -0.026405   0.006740 -0.006866      -0.011917   
Income (in K/month)    0.518550  -0.149095 -0.056477       0.466305   
ZIP Code              -0.008175  -0.024126  0.010120       0.002841   
Family members        -0.054160   0.022929  0.027610       0.013074   
CCAvg                  1.000000  -0.070689 -0.018175       0.328350   
Education             -0.070689   1.000000  0.010542       0.097033   
Mortgage              -0.018175   0.010542  1.000000      -0.027384   
Personal Loan          0.328350   0.097033 -0.027384       1.000000   
Securities Account     0.020410  -0.015682 -0.009023       0.016977   
CD Account             0.082784  -0.006481  0.009639       0.245397   
Online                -0.021219  -0.024194 -0.010379      -0.015612   
CreditCard            -0.002066  -0.017171 -0.001902      -0.005567   

                       Securities Account  CD Account    Online  CreditCard  
ID                              -0.019406   -0.008217 -0.008090    0.016780  
Age (in years)                   0.005677    0.005973  0.018553    0.011078  
Experience (in years)            0.004286    0.008487  0.018213    0.012537  
Income (in K/month)              0.001725    0.114803  0.004190    0.004247  
ZIP Code                         0.000313    0.016307  0.023776    0.002531  
Family members                   0.022259   -0.000306  0.011020    0.000842  
CCAvg                            0.020410    0.082784 -0.021219   -0.002066  
Education                       -0.015682   -0.006481 -0.024194   -0.017171  
Mortgage                        -0.009023    0.009639 -0.010379   -0.

Better than understanding the correlation values above and then applying limited knowledge to select independent variables, we shall use the automatic feature selection concept in Machine Learning to automate the task of selecting important features to analyse. The rationale behind this approach is to automate and also to improve the accuracy to selecting important features which can be effective candidates for our models to train. Note that we will be automating model selection on the independent variables and our dependent variable ('Personal Loan') shall remain separate.

In [143]:
from sklearn.feature_selection import SelectFromModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

X = capDf.iloc[:, capDf.columns != 'Personal Loan']
Y = capDf['Personal Loan']

test_size = 0.20 # taking 70:30 training and test set
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=test_size)

We shall apply the Min Max Scaler on our data for allowing better analysis and for bringing the data to better homogenity for our models to train.

In [144]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [145]:
select = SelectFromModel(
    RandomForestClassifier(n_estimators = 100, random_state = 42), threshold = 'median')
select.fit(X_train_scaled, y_train)
X_train_l1 = select.transform(X_train_scaled)
X_test_l1 = select.transform(X_test_scaled)
print("X_train.shape: {}".format(X_train_scaled.shape))
print("X_train_l1.shape: {}".format(X_train_l1.shape))
print("X_test_l1.shape: {}".format(X_test_l1.shape))

X_train.shape: (3518, 13)
X_train_l1.shape: (3518, 7)
X_test_l1.shape: (880, 7)


As we see above, of the 13 independent variables, 7 have been selected by the Model Based Feature Selector as effective candidates for our models to train with.

In [146]:
from sklearn.linear_model import LogisticRegression
logmodel = LogisticRegression()
logmodel.fit(X_train_l1,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [147]:
from sklearn.metrics import classification_report
predictions = logmodel.predict(X_test_l1)
print(classification_report(y_test,predictions))

             precision    recall  f1-score   support

          0       0.96      1.00      0.98       817
          1       0.91      0.46      0.61        63

avg / total       0.96      0.96      0.95       880



In [148]:
logmodel.score(X_test_l1, y_test)

0.9579545454545455

Here, we see that our model has scored 95% which is a healthy model score.

In [149]:
# calculate accuracy measures and confusion matrix
from sklearn import metrics
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score, precision_score, auc

In [150]:
metrics.confusion_matrix(y_test, predictions)

array([[814,   3],
       [ 34,  29]])

In [151]:
pd.DataFrame(data=[accuracy_score(y_test, predictions), 
                   recall_score(y_test, predictions),
                   precision_score(y_test, predictions)], 
             index=["accuracy", "recall", "precision"])

0
accuracy   0.957955
recall     0.460317
precision  0.906250

Here, we see that although the precision and accuracy of our model are high, the recall is low.

In [152]:
from sklearn.neighbors import KNeighborsClassifier

MSE = {};
MSE_k = 0;
min_k = 1;
K = range(1,19)
for k in K:
    NNH = KNeighborsClassifier(n_neighbors= k)
    NNH.fit(X_train_l1, y_train.values.ravel())
    predicted_labels = NNH.predict(X_test_l1)
    MSE_k = 1 - accuracy_score(y_test, predicted_labels)
    MSE["{0}".format(k)]="{0}".format(MSE_k)
    
print(MSE)
print("Optimal Number of Neighbours and Corresponding MSE: {0}".format(min(MSE.items(), key=lambda x: x[1])))

{'1': '0.02386363636363631', '2': '0.03295454545454546', '3': '0.027272727272727226', '4': '0.035227272727272774', '5': '0.035227272727272774', '6': '0.0420454545454545', '7': '0.036363636363636376', '8': '0.040909090909090895', '9': '0.040909090909090895', '10': '0.0420454545454545', '11': '0.03863636363636369', '12': '0.04318181818181821', '13': '0.0420454545454545', '14': '0.04318181818181821', '15': '0.04318181818181821', '16': '0.045454545454545414', '17': '0.04431818181818181', '18': '0.045454545454545414'}
Optimal Number of Neighbours and Corresponding MSE: ('1', '0.02386363636363631')


Based on the above iterative processing to determine the optimum number of neighbours or 'k', we realize that the optimal number of neighbours which has the least corresponding Misclassification Error, is 1. Let us now check the accuracy for this model with k as 1.

In [153]:
# Call Nearest Neighbour algorithm
NNH = KNeighborsClassifier(n_neighbors= 1)
NNH.fit(X_train_l1, y_train.values.ravel())
predicted_labels = NNH.predict(X_test_l1)
pd.DataFrame(data=[accuracy_score(y_test, predicted_labels), 
                   recall_score(y_test, predicted_labels),
                   precision_score(y_test, predicted_labels)], 
             index=["accuracy", "recall", "precision"])

0
accuracy   0.976136
recall     0.698413
precision  0.956522

We see above that not only the recall has improved significantly from around 46% to about 70% after utilizing the K Nearest Neighours algorithm, but also the accuracy has improved from about 95.7% to over 97%, and also the precision has increased from 90.6% to 95.6%.

In [155]:

metrics.confusion_matrix(y_test, predicted_labels)

array([[815,   2],
       [ 19,  44]])

Above, we see the confusion matrix after applying the KNN model, where the increase in recall rate is starkly evident, as compared to applying the Logistic Regression Model.